README.md
<p align="center">
  <img src="https://img.alicdn.com/tfs/TB17UtINiLaK1RjSZFxXXamPFXa-606-643.png" alt="logo" width="20%"/>
</p>
<h1 align= "center">
FormRender
</h1>

<p align="center">
  <a href="https://www.npmjs.com/package/form-render?_blank">
    <img alt="npm" src="https://img.shields.io/npm/v/form-render.svg?maxAge=3600&style=flat-square"></a>
  <a href="https://github.com/alibaba/form-render/commits/dev">
    <img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/alibaba/form-render.svg?style=flat-square"></a>
  <a href="https://github.com/alibaba/form-render">
    <img alt="GitHub repo size" src="https://img.shields.io/github/repo-size/alibaba/form-render"></a>
  <a href="https://github.com/alibaba/form-render/issues?utf8=%E2%9C%93&q=">
    <img alt = "GitHub closed issues" src="https://img.shields.io/github/issues-closed/alibaba/form-render.svg?style=flat-square"></a>
  <a href="https://npmjs.org/package/form-render">
    <img alt = "NPM downloads" src="https://img.shields.io/npm/dm/form-render.svg?style=flat-square"></a>
  <a href="https://npmjs.org/package/form-render">
    <img alt = "NPM all downloads" src="https://img.shields.io/npm/dt/form-render.svg?style=flat-square"></a>
  <a>
    <img alt = "PRs Welcome" src="https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square"></a>
</p>

> 通过 JSON Schema 生成标准 Form，常用于自定义搭建配置界面生成

## 了解

- <a href="/docs/guide/design.md">设计理念</a>
- <a href="https://x-render.gitee.io/schema-generator/" target="_blank">schema 编辑器</a>
- <a href="https://x-render.gitee.io/form-render/playground" target="_blank">Playground</a> / <a href="https://codesandbox.io/s/form-renderjichudemo-8k1l5" target="_blank">Code Sandbox</a>
- <a href="/docs/guide/others/usedBy.md">常见场景</a>
- <a href="https://github.com/alibaba/form-render/projects/2" target="_blank">后期规划</a>

## 效果

<img src="https://gw.alipayobjects.com/mdn/feizhu_pla/afts/img/A*wyH4Rq-EqwQAAAAAAAAAAABkARQnAQ?raw=true" width="750px"/>

## 优势

<table><tr><td><img src="https://gw.alicdn.com/tfs/TB1f5TAJND1gK0jSZFsXXbldVXa-2152-1162.gif?raw=true" width="750px"/></td><td><img src="https://gw.alipayobjects.com/zos/demos_resources/5ay8d5/fr-flow-short-par2.gif?raw=true" width='750px' /></td></tr></table>

- 如上图，使用 [Schema 编辑器](https://x-render.gitee.io/schema-generator/) 快速生成可实现低上手成本、快速搭建
- 支持 Ant Design 和 Fusion Design 主流的视觉主题
- 使用 JSON Schema 标准协议描述表单配置，并搭配丰富类型且可扩展的组件
- 支持 1 排 N、横纵排、支持对象无限嵌套、自定义正则校验、自定义样式组件、列表拖拽等特性
- 已在阿里云、淘宝、天猫、飞猪、亚博科技、安全智能、新零售行业工作台、人工智能实验室等多 BU 多场景使用，简单使用同时支持复杂场景使用
- 使用上有详细文档，维护上有专人支持

## 安装

```shell
npm i form-render
# or
yarn add form-render
```

同时安装依赖的组件库 (依赖哪个装哪个，天然支持 antd 和 fusion 两套。其他的组件库可通过 widgets 的方式传入，见文档 - 高级功能 - 自定义组件)

```shell
# 例如
yarn add antd
# 或者
yarn add @alifd/next
```

注：一般来说，想使用 form-render 的 antd 主题的项目，dependencies 里肯定装了 antd。这里强调一下这步，主要是针对想尝试写个 demo 玩玩的同学，记得要安装对应的 ui 组件库

## 快速使用

```jsx
import React, { useState } from 'react';
// 使用 Ant Design 风格
import FormRender from 'form-render/lib/antd';
// 使用 Fusion 风格
// import FormRender from 'form-render/lib/fusion';
// import '@alifd/next/dist/next.min.css';

const schema = {
  type: 'object',
  properties: {
    string: {
      title: '字符串',
      type: 'string',
      'ui:disabled': true,
    },
    select: {
      title: '单选',
      type: 'string',
      enum: ['a', 'b', 'c'],
      enumNames: ['早', '中', '晚'],
      'ui:width': '50%', // uiSchema 合并到 schema 中（推荐写法，书写便捷）
    },
  },
};

function Demo() {
  const [formData, setData] = useState({});
  const [valid, setValid] = useState([]);

  const onSubmit = () => {
    // valid 是校验判断的数组，valid 长度为 0 代表校验全部通过
    if (valid.length > 0) {
      alert(`校验未通过字段：${valid.toString()}`);
    } else {
      alert(JSON.stringify(formData, null, 2));
    }
  };

  return (
    <div style={{ maxWidth: 600 }}>
      <FormRender
        schema={schema}
        formData={formData}
        onChange={setData}
        onValidate={setValid}
        displayType="row" // 详细配置见下
      />
      <button onClick={onSubmit}>提交</button>
    </div>
  );
}

export default Demo;
```

### API

| Prop             |        Type         | Required | Default  |                                  Description                                  |
| ---------------- | :-----------------: | :------: | :------: | :---------------------------------------------------------------------------: |
| **schema**       |      `Object`       |    ✓     |    {}    |                  详见 [schema 配置](/docs/config/schema.md)                   |
| **uiSchema**     |      `Object`       |          |    {}    | 详见 [uiSchema 配置](/docs/config/uiSchema.md)（**一般建议合并到 `schema`**） |
| **formData**     |      `Object`       |          |    {}    |                                 配置表单数据                                  |
| **onChange**     |     `Function`      |    ✓     | () => {} |                               数据更改回调函数                                |
| **onValidate**   |     `Function`      |          | () => {} |                               表单输入校验回调                                |
| **displayType**  |      `String`       |          |  column  |                  设置表单横向排列或者纵向排序`column`/`row`                   |
| **showDescIcon** |      `Boolean`      |          |  false   |        描述是否用 tooltip 展示。`displayType`为 `row`时建议设为 `true`        |
| **readOnly**     |      `Boolean`      |          |  false   |                              预览模式/可编辑模式                              |
| **labelWidth**   | `Number` / `String` |          |   110    |    全局设置 label 长度(默认 110)。数字值单位为 px，也可使用'20%'/'2rem'等     |
| **widgets**      |      `Object`       |          |    {}    |                                  自定义组件                                   |

**注 1：** 设置表单 `displayType` 为 row 时候，请设置 `showDescIcon` 为 `true`，隐藏说明，效果会更好

**注 2：** **onChange** 方法会用于初始化表单 data，如果不写会造成没有初始值的表单元素无法渲染（出现不报错也不显示的情况）

**注 3：** FormRender 默认布局会占满它的父级元素，建议用一个`div`包裹 FormRender 用于表单布局样式调整

### 不常用 API

| Prop             |    Type    |   usage   |  Default  |                               Description                                |
| ---------------- | :--------: | :-------: | :-------: | :----------------------------------------------------------------------: |
| **mapping**      |  `Object`  | sometimes | undefined | 用于修改默认组件映射表，一般用于让自定义组件作为默认选择(详见自定义组件) |
| **column**       |  `Number`  | sometimes |     1     |          **整体**布局 1 排 N，局部的 1 排 N 一般使用`ui:width`           |
| **useLogger**    | `Boolean`  |   debug   |   false   |      当 useLogger 为 true 时，会在 console 展示所有的 formData 变化      |
| **name**         |  `String`  | very rare |  \$form   |                                表单的名称                                |
| **showValidate** | `Boolean`  | very rare |   true    |                             是否展示校验信息                             |
| **onMount**      | `Function` | very rare | undefined |                    onMount 在 formData 首次计算后触发                    |

[详见“不常用 props”](/docs/config/props2.md)

## 更多使用

- vscode 插件：快速准确书写 schema 一直是使用者的痛点，为此我们准备了 schema 书写利器： `form-render snippets`（vscode 插件），在 vscode 商店输入 ‘formrender’

- 支持 TypeScript：详见[如何在 TypeScript 项目中使用](https://x-render.gitee.io/form-render/guide/others/faq)

- **支持 Ant Design 自定义主题不被覆盖**

  - 安装 webpack 插件
    ```bash
    npm install webpack-plugin-fr-theme  --save-dev
    ```
  - 配置 webpack.config.js 文件

    ```js
    const WebpackPluginFrTheme = require('webpack-plugin-fr-theme');

    {
        ...
        plugins: [
          new WebpackPluginFrTheme(),
        ],
        ...
    }
    ```

## 调试

```shell
> git clone https://github.com/alibaba/form-render.git
> npm i
> npm start
```

## 支持

- 在公司或个人项目中使用 FormRender，关注 <a href="/CHANGELOG.md" target="_blank">Changelog</a>
- 如果你觉得 FormRender 还不错，可以通过 Star 来表示你的喜欢
- 加入钉钉聊天群帮忙解答使用问题

<img src="https://img.alicdn.com/tfs/TB1CprJg5pE_u4jSZKbXXbCUVXa-894-1087.jpg" width="300px" />

## 贡献

想贡献代码、解 BUG 或者提高文档可读性？非常欢迎一起参与进来，在提交 MR 前阅读一下 [Contributing Guide](https://github.com/alibaba/form-render/blob/master/CONTRIBUTING.md)

感谢给 FormRender 贡献代码的你们，以及 JetBrains 提供 Free 使用！

<a href="https://github.com/alibaba/form-render/graphs/contributors"><img src="https://opencollective.com/form-render/contributors.svg?width=890&button=false"/></a><a href="https://www.jetbrains.com/?from=form-render"><img src="https://img.alicdn.com/tfs/TB1gPDDJKL2gK0jSZFmXXc7iXXa-2000-2168.png" width="100px" /></a>

## 协议

- 遵循 MIT 协议
- 请自由地享受和参与开源


CHANGELOG.md
# Change Log

### 0.9.2

- [!] `format: url` 类型使用的兼容判断放宽，考虑到一些非常见的的网址

### 0.9.1

此版本主题是校验的扩展

- [+] 新增了 format：`email`/`url`有全新的展示和校验, `image`也添加了数据校验
- [!] type: string 类型的输入框，现在会校验是否在输入内容前后有空格。如果不希望空格被校验，可以使用 { noTrim: true } 属性

### 0.9.0

- [+] 【Breaking】通过 ref 的方式将内部方法暴露出来，目前只有 `resetData`，具体使用参见 [api](https://x-render.gitee.io/form-render/config/api)
  如果有用户在之前因为任何原因在 form-render 上挂了 ref，那这个是 breaking change(虽然感觉情况不多见)
- [!] 修改了 onMount 的逻辑，现在 onMount 会在 formData 首次计算后触发
- [!] 【Breaking】自定义组件接受到的入参 `readonly` 改为 `readOnly` 遵守惯例的驼峰格式。同时修复了一个 readOnly 无效的问题
- [+] 对象类型接受 `collapsed` 入参，用于折叠

### 0.8.11

- [+] 表单 props `column`、`displayType`、`showDescIcon` 属性都有了在 schema 中的对应的属性`ui:column`、`ui:displayType`、`ui:showDescIcon`, 可以在任何一层局部使用。同时意味着现在的 schema 已经能够完整传递所有展示信息（而不需要写一些 props）
  <img src="https://img.alicdn.com/tfs/TB1oz5ervzO3e4jSZFxXXaP_FXa-882-348.png" width="250" />
- [!] 列表点击“新增”会自动翻页, 列表修改 pageSize 能正常生效

### 0.8.9 - 0.8.10

- [+] 数组组件支持自动分页，并添加 `ui:options/pageSize` 可用于控制每页显示数量 ([#201](https://github.com/alibaba/form-render/issues/201))
- [+] Array 类型组件兼容标准的 JSON schema 写法 ([@siyi98](https://github.com/siyi98) in [#191](https://github.com/alibaba/form-render/pull/191))
- [!] 修复了两个 react dependencies 相关的问题，项目现在能正常支持所有 react 16.8 以上的版本（因为内部用到了 hooks）。([#197](https://github.com/alibaba/form-render/issues/197))
- [!] 真正不再加载整个 `@ant-design/icons`，大量减少加载体积 ([@F-loat](https://github.com/F-loat) in [#202](https://github.com/alibaba/form-render/pull/202))
- [!] 避免 enum 和 enumNames 乱写会造成整个项目 crash
- [!] 修复文档，各种链接跳转正确地址，fusion demo 缺失 css 的问题解决 ([@siyi98](https://github.com/siyi98) in [#191](https://github.com/alibaba/form-render/pull/191))

### 0.8.8

感谢各位的 PR。双十一有点忙，之后 FR 的发布会稳定进入一周一版

- [!] 确保 color 组件在 antd 展示 100% 伸缩
- [!] 添加更多字段允许函数表达式('minimum','maximum', 'minLength', 'maxLength') ([@monijiang](https://github.com/monijiang) in [#177](https://github.com/alibaba/form-render/pull/177))
- [!] maximum 字段支持浮点数 ([@fayeah](https://github.com/fayeah) for [#188](https://github.com/alibaba/form-render/issues/188))
- [!] 时间和时间范围组件都是非受控组件造成初始数据无法展示的问题 ([@zhangchen915](https://github.com/zhangchen915) in [#177](https://github.com/alibaba/form-render/pull/172))
- [!] 支持离线使用，将 cdn 资源图片替换为 base64 ([#158](https://github.com/alibaba/form-render/issues/158))
- [!] Demo 和文档的链接以及 schema 写法的翻新
- [!] 列表折叠时，多选框内容正确显示（之前只展示{复杂结构}）([@terryz95](https://github.com/terryz95) in [#184](https://github.com/alibaba/form-render/pull/184))

### 0.8.7

- [!] 校验 pattern 的逻辑回滚到之前的模式。只有填写了值之后才会去校验 pattern 的正则表达式
- [!] 修复了 switch 和时间区间组件有默认值时不展示的问题 ([#169](https://github.com/alibaba/form-render/issues/169))
- [!] 修复了外部人为修改 schema 时数据不更新的的问题 ([#173](https://github.com/alibaba/form-render/issues/173))
- [!] 修复了选择组件在 js 文件里使用 `default: undefined` 无效的问题 ([#174](https://github.com/alibaba/form-render/issues/174))

### 0.8.6

- [!] 优化不再加载整个`@ant-design/icons`。暂时去掉 antd 侧使用 string 名称来描述列表新增组件的功能
- [!] 重写 antd 的 upload 组件，去掉一些内置配置，允许在`ui:options`里自由配置
- [!] 解决 fusion 的样式覆盖问题 ([#165](https://github.com/alibaba/form-render/issues/165))
- [!] 表单类型为 number 输入值为 0 时没有触发 pattern 校验 ([#150](https://github.com/alibaba/form-render/issues/150))
- [!] boolean 类设置必填时，选择 false 时必填校验不通过 ([#162](https://github.com/alibaba/form-render/issues/162))
- [!] 文档更新，新功能补充说明，tooltip 太模糊换了个图标，钉钉群二维码更新 ([#163](https://github.com/alibaba/form-render/issues/163))

### 0.8.5

- [!] fix 了一个 dependencies 的问题

### 0.8.4

下面几个版本的一个大主题会是性能的提升

- [+] 添加了一个实验性的 `path` 入参，填写 url 后 form-render 会自动抓取云端的 schema 进行渲染 （由于是实验性的，暂时不在文档中放出，等稳定后再给说明）
- [!] 节流了内部校验逻辑，避免每次表单操作都调用

### 0.8.3

- [!] 避免了有些情况表单内部输入造成的反复渲染。这个渲染机制是为了监听外部人为的改动数据，现在内部输入不再会触发
- [!] 避免了外部硬注入非 string 类的数据造成 input 的校验报 error
- [!] 避免了外部硬注入非 string 类的数据造成 input 的 maxLength 的展示错误
- [!] 修复了`ui:hidden`为函数表达式时，校验无法准确判断的问题
- [!] 修复了单选组件 `default: null` 时的报错

### 0.8.2

- [!] 优化新的校验机制的触发，防止庞大的 schema 反复触发
- [!] 优化列表的展示，删除按钮上移，现在的 list 占位更小

### 0.8.1

- [+] 单选现在允许设置 default: null，使得默认不选择任何一个选项 ([#146](https://github.com/alibaba/form-render/issues/146))
- [!] "ui:options" 里每个 key 如果是函数表达式，也能正常解析 ([#104](https://github.com/alibaba/form-render/issues/104))
- [!] uniqueItems 的判断，支持到复杂的 item 结构（详见 [“新功能”](https://x-render.gitee.io/form-render/guide/new-feature)）
- [!] 修复了一个不常见的新校验机制和历史写法（现在的函数表达式的前身）产生的 bug

### 0.8.0

- [Breaking] 对`antd`和`@alifd/next`的依赖修改为`peerDependencies`。意味着这两个包需要使用方提供。这么做避免了之前使用 form-render 时会打入两个 antd（或 fusion）的问题。目前的使用场景看来，90%以上的用户都已安装了 `antd` 或者 `@alifd/next`。所以这个升级对于 90% 的用户理应是无缝的。对于独立使用 form-render 而未安装任何组件的同学，需要 `npm i antd` 一下。详见文档的 [开始使用](https://x-render.gitee.io/form-render/guide/getting-started)
- [!] 解决了首次渲染时就会展示校验红字的问题，现在通过设定`showValidate`为 false 来关闭此行为，只有动过的 field 会提示校验，提交时通过将 showValidate 设为 true 来唤起全部校验信息的展示
- [!] 解决了 antd 组件 textArea 和 input 人为传入 undefined 和 null 为值时的报错 ([#138](https://github.com/alibaba/form-render/issues/138))
- [!] description 隐藏时采用 info 图标代替问号图标，符合业界规范 ([#133](https://github.com/alibaba/form-render/issues/133))

### 0.7.3

- [!] 文档链接修复
- [!] playground 彻底全屏展示，且修复 z-index 的问题

### 0.7.2

- [+] input/textarea 有 maxLength 属性时，展示字数长度的提示 ([#127](https://github.com/alibaba/form-render/issues/127))
- [+] `"ui:options"/buttons` 支持 组件库的 button props 扩展 (原先的 `ui:extraButtons` 已经弃用，遵循 FR 的 schema 规范，见“设计理念”)。在按钮较多场景时，用户希望有重点的弱化或者强化部分按钮 ([#122](https://github.com/alibaba/form-render/issues/122))
- [!] 校验失败的文字框的 box-shadow 展示正确淡红色（而不是以前的浅蓝）
- [!] 当多选 checkbox 折行时，保持展示对齐（antd/fusion）([#125](https://github.com/alibaba/form-render/issues/125))

### 0.7.0

- [!] 文档大幅度翻新，构建发布流切换到 dumi、father
- [!] 使用 schema 字段替换 propsSchema 字段，因为后者太长且意义不明（目前保持兼容）

### 0.6.2

- [!] fusion 侧再也不需用户手动引入 css。~~`import "@alifd/next/dist/next.min.css";`~~

### 0.6.1

- [!] ui:button 的 callback 的入参从（value, onChange, newItem）变为 （value, onChange, schema, newItem）
- [!] modal/drawer list 类数据，在配置按钮旁显示（xx 条数据）
- [!] modal 添加确定按钮，给使用者一个配置成功的感受
- [!] 修复了 fusion 下 modal 的 2 个展示瑕疵
- [!] 在“新功能”添加新的 modal 样例

### 0.6.0

- [!] 解决 css 覆盖问题

.umirc.js
import { defineConfig } from 'dumi';
import path from 'path';

export default defineConfig({
  title: 'form-render',
  favicon: 'https://img.alicdn.com/tfs/TB17UtINiLaK1RjSZFxXXamPFXa-606-643.png',
  logo: 'https://img.alicdn.com/tfs/TB17UtINiLaK1RjSZFxXXamPFXa-606-643.png',
  outputPath: 'docs-dist',
  dynamicImport: {},
  // menus: {
  //   '/guide': [
  //     {
  //       title: '介绍',
  //       children: ['guide/getting-started', 'guide/new-feature'],
  //     },
  //   ],
  // },
  navs: [
    null,
    {
      title: '表单设计器',
      path: 'https://x-render.gitee.io/schema-generator/',
    },
    {
      title: 'table-render',
      path: 'https://x-render.gitee.io/table-render/',
    },
    { title: 'GitHub', path: 'https://github.com/alibaba/form-render' },
    {
      title: '更新日志',
      path: 'https://github.com/alibaba/form-render/blob/master/CHANGELOG.md',
    },
    {
      title: '国内镜像',
      path: 'https://x-render.gitee.io/form-render/',
    },
  ],
  mode: 'site',
  alias: { 'form-render/dist': path.resolve(__dirname, 'dist') },
  base: '/form-render/',
  publicPath: '/form-render/',
  exportStatic: {},
  extraBabelPlugins: [
    [
      'import',
      {
        libraryName: 'antd',
        libraryDirectory: 'lib',
        style: 'css',
      },
      'antd',
    ],
    [
      'import',
      {
        libraryName: '@alifd/next',
        libraryDirectory: 'lib',
      },
      '@alifd/next',
    ],
  ],
  // more config: https://d.umijs.org/config
});
